ARROW-7705: [Rust] Initial sort implementation

nevi-me opened the pull request:
This implements `sort_to_indices` which returns an u32 array of indices that results in the array being sorted. It then implements `sort` by using `take` to sort the array.

Similar to the approach taken in CPP, I think we can start with a correct approach, then focus on performance after. I will keep this PR a draft for comments and until I complete the initial sort implementation (including sorting in descending order).

nevi-me commented:
I'm still researching around null-order semantics, and want to complete this PR once I've found a working solution.

I'm so far thinking of creating the following:

```rust
pub struct SortOptions {
    pub descending: bool, // defaults to false
    pub nulls_first: bool, // defaults to false (current behaviour on the draft PR)
}
```

Follow-up PRs can then add sorting for types that the default sort doesn't handle, such as:
* lists,
* structs, to allow for multiple key sort (probably the most difficult)

nevi-me commented:
This is ready for review @paddyhoran @sunchao @andygrove. I'd appreciate some inputs/opinions on how we're handling descending sort. Not sure if `std::cmp;:Reverse` is better than sorting and reversing the whole array of indices.

andygrove commented on rust/arrow/src/compute/kernels/sort.rs:
```diff
@@ -0,0 +1,295 @@
+        DataType::UInt16 => sort_primitive::<UInt16Type>(values, v, n, &options),
+        DataType::UInt32 => sort_primitive::<UInt32Type>(values, v, n, &options),
+        DataType::UInt64 => sort_primitive::<UInt64Type>(values, v, n, &options),
+        // DataType::Float32 => sort_primitive::<Float32Type>(values, v, n, descending),
```
could you add a comment explaining why the floating point ones are commented out?

nevi-me replied:
I think I was trigger-happy when thinking that this was ready for review. Floats can't be sorted in the default way, so I need `ordered_float` or an equivalent. I left floats commented out because I wanted to investigate this.

I have at least 2 options, depending on which you prefer;

1. leave a note and open a JIRA to support floats
2. keep this PR open for longer so I can add the float support.

I'll try the latter in the interim.

andygrove replied:
Having a separate JIRA for sorting floats sounds good to me.

nevi-me replied:
https://issues.apache.org/jira/browse/ARROW-7924

andygrove commented on rust/arrow/src/compute/kernels/sort.rs:
```diff
@@ -0,0 +1,295 @@
+    T: ArrowPrimitiveType,
+    T::Native: std::cmp::Ord,
+{
+    let values = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
```
Can we use `expect` instead of `unwrap`?

andygrove commented on rust/arrow/src/compute/kernels/sort.rs:
```diff
@@ -0,0 +1,295 @@
+        .map(|index| (index as u32, values.value(index)))
+        .collect::<Vec<(u32, T::Native)>>();
+    let mut nulls = null_indices;
+    if !options.descending {
```
I think this would be more readable if the if were inverted to remove the `!` from the condition?

```rust
if options.decending {
} else {
}
```

andygrove replied:
In fact, I think I would prefer the options to have a field named `ascending` rather than `descending` since ascending is typically the default.

nevi-me replied:
My justification for using `descending` instead of `ascending` is that the default behaviour should not require passing options. As boolean's default is false, I'd rather the caller opt in explicitly to sorting in a descending order.
I could set the default to ascending, but I didn't like the below as it'd be less obvious:

```rust
impl Default for SortOptions {
    fn default() -> Self {
        Self {
            ascending: true,
            nulls_first: false,
        }
    }
}
```

nevi-me commented:
I haven't had time to complete this, I got stuck while writing a macro to help create test cases. I'll ask for help online when I get enough time. I'd like to complete sort capabilities before 1.0.0

nevi-me commented:
@andygrove I've addressed the review comments, and have added tests.
I tried creating a macro that allows me to reuse the same inputs for different array types, but wasn't able to, so the tests are rather repetitive.

PTAL @paddyhoran 

I'll work on `sort` support for other array types, with the aim of completing the work before `1.0.0`

paddyhoran commented:
@nevi-me this needs a rebase now.  Once you do that, I'll take a look so we can get this merged.

paddyhoran commented on rust/arrow/src/compute/kernels/sort.rs:
```diff
@@ -0,0 +1,671 @@
+/// Performs a stable sort on values and indices, returning nulls after sorted valid values,
+/// while preserving the order of the nulls.
+///
+/// Returns and `ArrowError::ComputeError(String)` if the array type is either unsupported by `sort_to_indices` or `take`.
```
nit: and -> an

nevi-me replied:
Thanks, I'll merge this after CI completes

nevi-me commented on rust/arrow/src/compute/kernels/sort.rs:
```diff
@@ -0,0 +1,671 @@
+/// Performs a stable sort on values and indices, returning nulls after sorted valid values,
+/// while preserving the order of the nulls.
+///
+/// Returns and `ArrowError::ComputeError(String)` if the array type is either unsupported by `sort_to_indices` or `take`.
```
```suggestion
/// Returns an `ArrowError::ComputeError(String)` if the array type is either unsupported by `sort_to_indices` or `take`.
```
